File: fsharp/bank-account/BankAccount.fs
module BankAccount

open System
open System.Collections.Concurrent

let store = ConcurrentDictionary<Guid, decimal>()

let mkBankAccount() = Guid.NewGuid()

let openAccount (account: Guid) =
    store.AddOrUpdate(account, 0m, (fun _ _ -> 0m)) |> ignore
    account

let closeAccount (account: Guid) =
    store.TryRemove(account) |> ignore
    account

let getBalance (account: Guid) =
    let mutable balance = 0m
    if store.TryGetValue(account, &balance) then
        Some balance
    else
        None

let updateBalance change (account: Guid) =
    if store.ContainsKey(account) then
        store.AddOrUpdate(account, change, (fun _ balance -> balance + change)) |> ignore
    account

File: csharp/bank-account/BankAccount.cs
using System;

public class BankAccount
{
    private enum BankAccountState { Open, Closed }

    private static readonly object ConcurrencyLock = new object();

    private BankAccountState _state = BankAccountState.Closed;
    private decimal _balance;

    public decimal Balance
    {
        get
        {
            lock (ConcurrencyLock)
            {
                ValidateAccountIsOpen();
                return _balance;
            }
        }
    }

    public void Open()
    {
        lock (ConcurrencyLock)
        {
            _state = BankAccountState.Open;
        }
    }

    public void Close()
    {
        lock (ConcurrencyLock)
        {
            _state = BankAccountState.Closed;
        }
    }

    public void UpdateBalance(decimal change)
    {
        ValidateAccountIsOpen();
        lock (ConcurrencyLock)
        {
            _balance += change;
        }
    }

    private void ValidateAccountIsOpen()
    {
        if (_state != BankAccountState.Open)
        {
            throw new InvalidOperationException("Account is not open");
        }
    }
}